File: src/Backend/Core/DTO/CommunityDTO.py
from typing import Optional, List, Literal
from pydantic import BaseModel, Field
from datetime import datetime


# DTO for user search results
class UserSearchResultDTO(BaseModel):
    userId: int
    displayName: str
    level: Optional[int] = None
    avatarUrl: Optional[str] = None
    relationship: Literal["none", "friends", "incoming_request", "outgoing_request", "blocked"]


# DTO for friend list
class FriendListItemDTO(BaseModel):
    userId: int
    displayName: str
    level: Optional[int] = None
    avatarUrl: Optional[str] = None


# DTO for friend request
class FriendRequestItemDTO(BaseModel):
    requestId: int
    userId: int
    displayName: str
    level: Optional[int] = None
    avatarUrl: Optional[str] = None
    requestDate: Optional[datetime] = None


# DTO for friend requests response
class FriendRequestsResponseDTO(BaseModel):
    incoming: List[FriendRequestItemDTO]
    outgoing: List[FriendRequestItemDTO]


# DTO for creating friend request
class CreateFriendRequestDTO(BaseModel):
    toUserId: int = Field(..., gt=0)


# DTO for leaderboard entry
class LeaderboardEntryDTO(BaseModel):
    rank: int
    userId: int
    displayName: str
    xp: int
    level: Optional[int] = None
    avatarUrl: Optional[str] = None


# DTO for current user leaderboard position
class CurrentUserLeaderboardDTO(BaseModel):
    rank: int
    xp: int
    level: Optional[int] = None


# DTO for leaderboard response
class LeaderboardResponseDTO(BaseModel):
    scope: Literal["friends", "global"]
    period: Literal["week", "month"]
    entries: List[LeaderboardEntryDTO]
    currentUser: Optional[CurrentUserLeaderboardDTO] = None


# DTO for friend visit profile
class FriendVisitProfileDTO(BaseModel):
    userId: int
    displayName: str
    level: Optional[int] = None
    avatarUrl: Optional[str] = None
    avatarImageUrl: Optional[str] = None
    avatarPreviewUrl: Optional[str] = None
    homeSceneId: Optional[str] = None
    homePreviewUrl: Optional[str] = None
    bio: Optional[str] = None
    totalExercises: Optional[float] = None
    totalDistance: Optional[float] = None
    totalSteps: Optional[int] = None
    longestStreak: Optional[int] = None


# DTO for friend stats summary
class FriendStatsDTO(BaseModel):
    range: Literal["week", "month"]
    workoutsCount: int
    minutesTotal: int
    xpTotal: int


# DTO for friend recent workout item
class FriendRecentWorkoutDTO(BaseModel):
    id: int
    type: str
    startedAt: datetime
    durationMinutes: int
    xpEarned: Optional[int] = None
    calories: Optional[int] = None
    avgHr: Optional[int] = None
    exercisesCount: Optional[int] = None


# DTO for friend recent workouts response
class FriendRecentWorkoutsDTO(BaseModel):
    items: List[FriendRecentWorkoutDTO]


# DTO for user search query
class UserSearchQueryDTO(BaseModel):
    q: str = Field(..., min_length=1)


# DTO for leaderboard query
class LeaderboardQueryDTO(BaseModel):
    scope: Literal["friends", "global"] = "friends"
    period: Literal["week", "month"] = "week"


# DTO for friend dashboard summary - XP progress
class XpProgressDTO(BaseModel):
    currentXp: int
    nextLevelXp: int
    progress: float  # Percentage 0-100


# DTO for friend dashboard summary - Streak info
class StreakInfoDTO(BaseModel):
    days: int
    label: str  # e.g., "5 days streak" or "No streak"


# DTO for friend dashboard summary - This week stats
class ThisWeekStatsDTO(BaseModel):
    workoutsCount: int
    minutesTotal: int
    xpTotal: int


# DTO for friend dashboard summary response
class FriendDashboardSummaryDTO(BaseModel):
    userId: int
    displayName: str
    level: int
    avatarNickname: str
    avatarImageUrl: Optional[str] = None
    themeColor: Optional[str] = None
    xp: dict  # Will contain currentXp, nextLevelXp, progress
    streak: dict  # Will contain days, label
    thisWeek: dict  # Will contain workoutsCount, minutesTotal, xpTotal

File: src/TOMI/models/dto/Community.dto.ts
// Community-related DTOs for TOMI Frontend

export interface UserSearchResult {
  userId: number;
  displayName: string;
  level?: number;
  avatarUrl?: string;
  relationship: 'none' | 'friends' | 'incoming_request' | 'outgoing_request' | 'blocked';
}

export interface FriendListItem {
  userId: number;
  displayName: string;
  level?: number;
  avatarUrl?: string;
}

export interface FriendRequestItem {
  requestId: number;
  userId: number;
  displayName: string;
  level?: number;
  avatarUrl?: string;
  requestDate?: string;
}

export interface FriendRequestsResponse {
  incoming: FriendRequestItem[];
  outgoing: FriendRequestItem[];
}

export interface LeaderboardEntry {
  rank: number;
  userId: number;
  displayName: string;
  xp: number;
  level?: number;
  avatarUrl?: string;
}

export interface CurrentUserLeaderboard {
  rank: number;
  xp: number;
  level?: number;
}

export interface LeaderboardResponse {
  scope: 'friends' | 'global';
  period: 'week' | 'month';
  entries: LeaderboardEntry[];
  currentUser?: CurrentUserLeaderboard;
}

export interface FriendVisitProfile {
  userId: number;
  displayName: string;
  level?: number;
  avatarUrl?: string;
  avatarImageUrl?: string;
  avatarPreviewUrl?: string;
  homeSceneId?: string;
  homePreviewUrl?: string;
  bio?: string;
  totalExercises?: number;
  totalDistance?: number;
  totalSteps?: number;
  longestStreak?: number;
}

export interface FriendStats {
  range: 'week' | 'month';
  workoutsCount: number;
  minutesTotal: number;
  xpTotal: number;
}

export interface FriendRecentWorkout {
  id: number;
  type: string;
  startedAt: string;
  durationMinutes: number;
  xpEarned?: number;
  calories?: number;
  avgHr?: number;
  exercisesCount?: number;
}

export interface FriendRecentWorkoutsResponse {
  items: FriendRecentWorkout[];
}

export interface CreateFriendRequest {
  toUserId: number;
}

// Friend Dashboard Summary DTOs
export interface XpProgress {
  currentXp: number;
  nextLevelXp: number;
  progress: number; // Percentage 0-100
}

export interface StreakInfo {
  days: number;
  label: string; // e.g., "5 days streak" or "No streak"
}

export interface ThisWeekStats {
  workoutsCount: number;
  minutesTotal: number;
  xpTotal: number;
}

export interface FriendDashboardSummary {
  userId: number;
  displayName: string;
  level: number;
  avatarNickname: string;
  avatarImageUrl?: string;
  themeColor?: string;
  xp: XpProgress;
  streak: StreakInfo;
  thisWeek: ThisWeekStats;
}

